pandas/tests/reshape/test_concat.py

```diff
@@ -1771,6 +1772,148 @@ def test_concat_tz_series_tzlocal(self):
         assert result.dtype == 'datetime64[ns, tzlocal()]'
 
+    def test_concat_NaT_dataframes_all_NaT_axis_0(self):
+        # GH 12396
+        expected = pd.DataFrame([pd.NaT, pd.NaT, pd.NaT], index=[0, 1, 0])
+
+        # non-timezone aware
+        first = pd.DataFrame([[pd.NaT], [pd.NaT]])
+        second = pd.DataFrame([[pd.NaT]])
+
+        result = pd.concat([first, second], axis=0)
+        assert_frame_equal(result, expected, check_datetimelike_compat=True)
+
+        # one side timezone-aware
+        dtype = DatetimeTZDtype('ns', tz='UTC')
+        first = pd.DataFrame([[pd.NaT], [pd.NaT]], dtype=dtype)
+
+        result = pd.concat([first, second], axis=0)
+        # upcasts for mixed case
+        expected = expected.apply(lambda x: x.astype(dtype))
+        assert_frame_equal(result, expected, check_datetimelike_compat=True)
+
+        # both sides timezone-aware
+        second = pd.DataFrame([[pd.NaT]], dtype=dtype)
+
+        # upcasts to tz-aware
+        result = pd.concat([first, second], axis=0)
+        assert_frame_equal(result, expected, check_datetimelike_compat=True)
+
+    def test_concat_NaT_dataframes_all_NaT_axis_1(self):
+        # GH 12396
+        expected = pd.DataFrame([[pd.NaT, pd.NaT], [pd.NaT, pd.NaT]],
+                                columns=[0, 1])
+
+        # non-timezone aware
+        first = pd.DataFrame([[pd.NaT], [pd.NaT]])
+        second = pd.DataFrame([[pd.NaT]], columns=[1])
+
+        result = pd.concat([first, second], axis=1)
+        assert_frame_equal(result, expected, check_datetimelike_compat=True)
+
+        # one side timezone-aware
+        dtype = DatetimeTZDtype('ns', tz='UTC')
+        first = pd.DataFrame([[pd.NaT], [pd.NaT]], dtype=dtype)
+
+        # upcasts result to tz-aware
+        expected.loc[:, 0] = expected.loc[:, 0].astype(dtype)
+        result = pd.concat([first, second], axis=1)
+        assert_frame_equal(result, expected, check_datetimelike_compat=True)
+
+        # both sides timezone-aware
+        second = pd.DataFrame([[pd.NaT]], dtype=dtype, columns=[1])
+
+        # upcasts to tz-aware
+        expected = expected.apply(lambda x: x.astype(dtype))
+        result = pd.concat([first, second], axis=1)
+        assert_frame_equal(result, expected, check_datetimelike_compat=True)
+
+    def test_concat_NaT_dataframes_mixed_timestamps_and_NaT(self):
+        # GH 12396
+
+        # non-timezone aware
+        first = pd.DataFrame([[pd.NaT], [pd.NaT]])
+        second = pd.DataFrame([[pd.Timestamp('2015/01/01')],
+                               [pd.Timestamp('2016/01/01')]])
+
+        expected = pd.DataFrame([pd.NaT, pd.NaT, second.iloc[0, 0],
+                                second.iloc[1, 0]], index=[0, 1, 0, 1])
+
+        result = pd.concat([first, second], axis=0)
+        assert_frame_equal(result, expected, check_datetimelike_compat=True)
+
+        # one side timezone-aware
+        dtype = DatetimeTZDtype('ns', tz='UTC')
```

Comment:
I don't really like this use of DatetimeTZDtyep, you can just do an explict astype here